### PR TITLE
Add channelFromParams and preprocessInputs to workflowhelper

### PR DESCRIPTION
**Pull request by DriesSchaumont**

### CODE REVIEW

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 1]
```diff
@@ -610,3 +610,278 @@ def viashChannel(params, config) {
...
+      }
+      parValue = parValue.flatten()
+    }
+    [parName, parValue]
+  }
+}
+
+def _checkUniqueIds(List<Map> multiParam) {
+  /*
+   *  Check if the ids are unique acorss parameter sets
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:642]: typo

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 2]
```diff
@@ -610,3 +610,278 @@ def viashChannel(params, config) {
...
+      } else {
+          parValue = [ parValue ]
+      }
+      parValue = parValue.flatten()
+    }
+    [parName, parValue]
+  }
+}
+
+def _checkUniqueIds(List<Map> multiParam) {
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:640]: also call this function in preprocessParams?

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 3]
```diff
@@ -610,3 +610,278 @@ def viashChannel(params, config) {
...
+    if (isInputFile) {
+      if (parValue instanceof List) {
+        parValue = parValue.collect({path -> 
+          path !instanceof String ? path : file(getChild(relativeTo, path))
+        })
+      } else {
+        parValue = parValue !instanceof String ? path : file(getChild(relativeTo, parValue))
+      }
+    }
+  [parName, parValue]
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:670]: Indentation

**Review by rcannood — CHANGES_REQUESTED** [review]
Proposed approach:

- [ ] Check whether WorkflowHelper.nf works in OpenPipelines
- [ ] Redefine `paramToList` using the newly defined functions.
- [ ] Add a notice that `paramsToList` will be deprecated in Viash 0.7.0
- [ ] For Robrecht: Rename `src/test/scala/io/viash/platforms/NextFlowVdsl3PlatformInputFormatsTest.scala` to `src/test/scala/io/viash/nextflow/WorkflowHelperTest.scala`
- [ ] Create a new file `src/test/resources/testnextflowvdsl3/workflows/test_workflowhelper/main.nf` containing something like:
    ```groovyinclude { readConfig; viashChannel; helpMessage } from "${params.rootDir}/workflows/utils/WorkflowHelper.nf"
    config = readConfig("${params.rootDir}/workflows/pipeline3/config.vsh.yaml")
    
    workflow unittest1 {
      // create params object
      params = [foo: "bar"]
      config = [...]
      out = viashChannel(params, config)
      // todo: add checks
    }
    workflow unittest2 { ... }
    workflow unittest3 { ... }
    workflow unittest4 { ... }
    workflow unittest5 { ... }
    ```
    Which you can run using `nextflow run src/test/resources/testnextflowvdsl3/workflows/test_workflowhelper/main.nf -entry unittest1`.
  - [ ] For Robrecht or Hendrik: Add a test to the scala code which loops over the available unit tests and runs them.
In the meantime, you can already use this WorkflowHelper.nf in OpenPipelines to solve the issues we're currently having

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 4]
```diff
@@ -610,3 +610,278 @@ def viashChannel(params, config) {
...
 }
+
+def _splitParams(Map paramList, List<Map> multiArgumentsSettings){
+  /*
+   *   Split parameters for arguments that accept multiple values using their seperator
+   *
+   *   paramList: a Map containing parameters to split
+   *   multiArgumentsSettings: a List of the viash configuration argument entries 
+   *                           that have the property 'multiple: true'
+   */
```
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Done

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 1 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Fixed

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 3 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:678]: Done!

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 5]
```diff
@@ -610,3 +610,278 @@ def viashChannel(params, config) {
...
+
+def _checkUniqueIds(List<Map> multiParam) {
+  /*
+   *  Check if the ids are unique acorss parameter sets
+   */
+  def ppIds = multiParam.collect{it[0]}
+  assert ppIds.size() == ppIds.unique().size() : "All argument sets should have unique ids. Detected ids: $ppIds"
+}
+
+def _resolvePathsRelativeTo(Map paramList, List<Map<String, String>> inputFileSetttings, String relativeTo) {
```
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Works for some functions, but I cannot specify output type `Channel` for example

File: src/test/resources/testnextflowvdsl3/workflows/pipeline3/main.nf [hunk 6]
```diff
@@ -1,11 +1,13 @@
...
 workflow base {
   helpMessage(config)
 
-  viashChannel(params, config)
-    | view{"DEBUG: $it"}
+  viash_ch = channelFromParams(params, config)
+    | preprocessInputs(["config": config])
+    | view {"DEBUG: $it"}    
+
 }
```
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/test/resources/testnextflowvdsl3/workflows/pipeline3/main.nf]: Done!

File: CHANGELOG.md [hunk 7]
```diff
@@ -39,7 +45,7 @@
   For every release, there is now a date, title, and summary.
   This both improves the changelog itself but can then also be used to postprocess the CHANGELOG programmatically.
 
-* `VDSL3`: Add a default value for `id` when running a VDSL3 module as a standalone pipeline.
+* `VDSL3`: Add default to the anonymous workflow of a VDSL3 component without adding it to the functionality resources.
```
Comment (rcannood) [inline CHANGELOG.md:48]: please revert this change

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 8]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
 def paramsToChannel(params, config) {
+  System.err.println("Warning: paramsToChannel has deprecated in Viash 0.7.0. " +
+                     "Please use a combination of channelFromParams and preprocessInputs.")
   Channel.fromList(paramsToList(params, config))
 }
 
 def viashChannel(params, config) {
+  System.err.println("Warning: viashChannel has deprecated in Viash 0.7.0. " +
+                     "Please use a combination of channelFromParams and preprocessInputs.")
   paramsToChannel(params, config)
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:617]: VDSL3 components currently use `viashChannel()`, so it would be weird that the output of a `viash build -p nextflow` would generate deprecation warnings when we're the ones who created it ;) Can you change the VDSL3 helper so that it uses `channelFromParams()` and `preprocessInputs` instead?

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 9]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
 
 def paramsToChannel(params, config) {
+  System.err.println("Warning: paramsToChannel has deprecated in Viash 0.7.0. " +
+                     "Please use a combination of channelFromParams and preprocessInputs.")
   Channel.fromList(paramsToList(params, config))
 }
 
 def viashChannel(params, config) {
+  System.err.println("Warning: viashChannel has deprecated in Viash 0.7.0. " +
+                     "Please use a combination of channelFromParams and preprocessInputs.")
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:616]: You're throwing three deprecation warnings when running viashChannel. Can you try ensuring that the deprecation warning only gets printed once? Maybe something like:

```groovy
viashChannelDeprecationWarningPrinted = false

def viashChannel(params, config) {
  if (!viashChannelDeprecationWarningPrinted) {
    viashChannelDeprecationWarningPrinted = true
    System.err.println("Spiderpig, spiderpig!")
  }
  // ...
}
```

Not sure if this will work. Can you try it out?

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 10]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
   paramsToChannel(params, config)
     | map{tup -> [tup.id, tup]}
 }
+
+/**
+ * From a viash config, gather the config entries 
+ * for arguments that have property 'multiple: true'
+ */
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:624]: `/**` signifies javadoc, but a title and the `@param config` and `@return` entries are missing. Can you add them?

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 11]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+
+private List<Map> _getMultiParameterSettings(Map config) {
+  def configArguments = config.functionality.allArguments
+  def plainNameArguments = configArguments.findAll{it.containsKey("plainName")}
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:628]: why are you filtering based on whether it contains `plainName`? I think an entry in `.allArguments` should always have a `.plainName`?

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 12]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
   paramsToChannel(params, config)
     | map{tup -> [tup.id, tup]}
 }
+
+/**
+ * From a viash config, gather the config entries 
+ * for arguments that have property 'multiple: true'
+ */
+
+private List<Map> _getMultiParameterSettings(Map config) {
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:626]: Can this be removed? It seems like it's not being used.

If it is, I think `config.functionality.allArguments.findAll{it.multiple}` is more readable than `_getMultiParameterSettings(config)`, so I'd still propose removing it.

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 13]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+  def configArguments = config.functionality.allArguments
+  def plainNameArguments = configArguments.findAll{it.containsKey("plainName")}
+  return plainNameArguments.findAll({it.multiple})
+}
+
+/**
+ * Split parameters for arguments that accept multiple values using their seperator
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:633]: *separator

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 14]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+ *       the parameter accepts multiple values
+ *    - Cast the parameters to their corect types.
+ *    - Assertions:
+ *        ~ Check if any unknown parameters are found
+ * 
+ * @return The input parameters that have been processed.
+ */
+private Map<String, Object> applyConfigToOneParameterSet(Map<String, Object> paramValues, Map config){
+  def configArguments = config.functionality.allArguments
+  def plainNameArguments = configArguments.findAll{it.containsKey("plainName")}
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:838]: redundant

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 15]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+ * 
+ * @return A list of parameters with the first element of the event being
+ *         the event ID and the second element containing a map of the parsed parameters.
+ */
+ 
+private List<Tuple2<String, Map<String, Object>>> _paramsToParamSets(Map params, Map config){
+  /* Get different parameter types */
+  /*********************************/
+  def configArguments = config.functionality.allArguments // List[Map]
+  def plainNameArguments = configArguments.findAll{it.containsKey("plainName")}
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:912]: redundant

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 16]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+    if (parValue instanceof Collection && !parameterSettings ) {
+      assert parValue.size() == 1 : 
+      "Error: argument ${parName} has too many values.\n" +
+      "  Expected amount: 1. Found: ${parValue.size()}"
+      parValue = parValue[0]
+    }
+    [parName, parValue]
+  }
+  return parsedParamValues
+}
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:668]: This function will throw assertion errors on parameters in `parValues` that are not in a confi's allArguments if they are length > 1 or will turn flatten the list if length == 1. I think this is an unintended consequence.

Why not simply pass the whole `config.functionality.allArguments` into the function? As an added bonus, it will make the code readable as well:

```groovy
Map<String, Object> _splitParams(Map<String, Object> parValues, List<Map> allArguments){
  def parsedParamValues = parValues.collectEntries { parName, parValue ->
    arg = allArguments.find{it.plainName == parName}

    if (!arg) {
      // if argument is not found, do not alter
      return [parName, parValue]
    }

    // Check if parameter can accept multiple values.
    // If so, make sure parValue is a list
    if (arg.multiple) {
      if (parValue instanceof Collection) {
          parValue = parValue.collect{it instanceof String ? it.split(arg.multiple_sep) : it }
      } else if (parValue instanceof String) {
          parValue = parValue.split(arg.multiple_sep)
      } else if (parValue == null) {
          parValue = [] // should we leave parValue unchanged if it is null?
      } else {
          parValue = [ parValue ]
      }
      parValue = parValue.flatten()
    }

    // For all parameters check if multiple values are only passed for
    // arguments that allow it. Quietly simplify lists of length 1.
    if (!arg.multiple && parValue instanceof Collection) {
      assert parValue.size() == 1 : 
        "Error: argument ${parName} has too many values.\n" +
        "  Expected amount: 1. Found: ${parValue.size()}"
      parValue = parValue[0]
    }

    [parName, parValue]
  }

  return parsedParamValues
}
```

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 17]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+ * @param inputFileSetttings A Map of the viash configuration argument entries 
+ *                           that are considered input files (i.e. with 'direction: input'
+ *                           and 'type: file').
+ * @param relativeTo path of a file to resolve the parameters values to.
+ *
+ * @return A map of parameters where the location of the input file parameters have been resolved
+ *         resolved relatively to the provided path.
+ */
+private Map<String, Object> _resolvePathsRelativeTo(Map paramList, 
+  List<Map<String, String>> inputFileSetttings, String relativeTo) {
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:694]: *inputFileSettings

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 18]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+  paramList.collectEntries { parName, parValue ->
+    isInputFile = inputFileSetttings.find({it.plainName == parName})
+    if (isInputFile) {
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:697]: Same comment as before. I think its much more readable to pass all arguments to this function than to check whether its in a subset of input arguments.

Proposed alternative:

```groovy
private Map<String, Object> _resolvePathsRelativeTo(
  Map paramList,
  List<Map<String, Object>> arguments,
  String relativeTo
) {
  paramList.collectEntries { parName, parValue ->
    arg = arguments.find{it.plainName == parName}
    if (arg && arg.type == "file" && arg.direction == "input") {
      if (parValue instanceof Collection) {
        parValue = parValue.collect({path -> 
          path !instanceof String ? path : file(getChild(relativeTo, path))
        })
      } else {
        parValue = parValue !instanceof String ? path : file(getChild(relativeTo, parValue))
      }
    }
    [parName, parValue]
  }
}
```

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 19]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+ * @param params Input parameters from nextflow.
+ * @param multiArgumentsSettings A List of the viash configuration argument entries 
+ *                               that have the property 'multiple: true'.
+ * @param inputfileArguments A List of the viash configuration argument entries 
+ *                           that specify the input files.
+ *
+ * @return A list of parameter sets that were parsed from the 'param_list' argument value.
+ */
+private List<Tuple2<String, Map>> _parseParamListArguments(Map params, 
+  List<Map> multiArguments, List<Map> inputfileArguments){
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:723]: Now I'm even more convinced that you should just pass the config arguments rather than make different subsets of the same argument lists for downstream checks ;)

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 20]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+  }
+
+  // Reformat from List<Map> to List<Tuple2<String, Map>> by adding the ID as first element of a Tuple2
+  paramSets = paramSets.collect({ paramValues ->
+    [paramValues.get("id", null), paramValues.findAll{it.key != 'id'}]
+  })
+  // Split parameters with 'multiple: true'
+  paramSets = paramSets.collect({ paramSet ->
+    def id = paramSet[0]
+    def paramValues = paramSet[1]
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:763]: Can use `paramSets = paramSets.collect({ id, data ->` notation since you know exactly what will be in `paramSets` here

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 21]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+    def splitParamValues = _splitParams(paramValues, multiArguments)
+    [id, splitParamValues]
+  })
+  
+  // The paths of input files inside a param_list file may have been specified relatively to the
+  // location of the param_list file. These paths must be made absolute.
+  if (paramListFile){
+    paramSets = paramSets.collect({ paramSet ->
+      def id = paramSet[0]
+      def paramValues = paramSet[1]
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:773]: Can use `paramSets = paramSets.collect({ id, data ->` notation since you know exactly what will be in `paramSets` here

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 22]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+  return paramSets
+}
+
+/**
+ * Cast parameters to the correct type as defined in the Viash config.
+ */
+
+private Map<String, Object> _castParamTypes(Map<String, Object> parValues, Map config) {
+  def configArguments = config.functionality.allArguments
+    // Cast the input to the correct type according to viash config
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:788]: fix indentation

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 23]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+  def castParValues = parValues.collectEntries({ parName, parValue ->
+    paramSettings = configArguments.find({it.plainName == parName})
+    // dont parse parameters like publish_dir ( in which case paramSettings = null)
+    parType = paramSettings ? paramSettings.get("type", null) : null
+    if (! (parValue instanceof Collection)) {
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:793]: It's good you put parentheses around the instanceof statement because indeed Groovy will otherwise interpret it as `(!parValue) instanceof Collection` which is incorrect. If you want, you can use `if (parValue !instanceof Collection) {` instead.

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 24]
```diff
@@ -603,10 +606,467 @@ def paramsToList(params, config) {
...
+      parValue = parValue.collect{it as Boolean}
+    }
+
+    // simplify list to value if need be
+    if (paramSettings && !paramSettings.multiple) {
+      assert parValue.size() == 1 : 
+        "Error: argument ${parName} has too many values.\n" +
+        "  Expected amount: 1. Found: ${parValue.size()}"
+      parValue = parValue[0]
+    }
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:820]: This transformation is happening twice; here and in `_splitParams`. Is this intentional?

**Review by rcannood — CHANGES_REQUESTED** [review]
Nice to see some major improvements to the workflowhelper!

I have a few requests; see comments

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 24 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: I was needed before we applied `applyConfig` to everything, but not anymore.

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 11 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: I thought this was part of the old functionality, but it is not, I will remove the filtering

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 8 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:617]: Done, added to the VDSL3 helper

File: CHANGELOG.md (hunk 7 above)
Comment (DriesSchaumont) [reply CHANGELOG.md:52]: Done

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 12 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Done!

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 9 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: This works!

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 17 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Adjusted

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 18 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Adjusted

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 16 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Adjusted

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 25]
```diff
@@ -603,10 +610,447 @@ def paramsToList(params, config) {
...
+ * Apply the argument settings specified in a Viash config to a single parameter set.
+ *    - Split the parameter values according to their seperator if 
+ *       the parameter accepts multiple values
+ *    - Cast the parameters to their corect types.
+ *    - Assertions:
+ *        ~ Check if any unknown parameters are found
+ * 
+ * @return The input parameters that have been processed.
+ */
+private Map<String, Object> applyConfigToOneParameterSet(Map<String, Object> paramValues, Map config){
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:834]: This assumes paramValues has already been processed, right? (because kebab-case arguments will turn into camelCase and vice versa).

Can you add javadocs for the paramValues and config?

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf [hunk 26]
```diff
@@ -603,10 +610,447 @@ def paramsToList(params, config) {
...
+ * and the second element containg the the parameters themselves.
+ * Optional extra elements of the tuples will be passed to the output as is.
+ *
+ * @param args: A map that must contain a 'config' key that points
+ *              to a parsed config (see readConfig()). Optionally, a
+ *              'key' key can be provided which can be used to create a unique
+ *              name for the workflow process.
+ *
+ * @return: A workflow that allows processing a channel of Vdsl3 formatted events
+ * and apply a Viash config to them.
```
Comment (rcannood) [inline src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:1033]: No colon needed after params and args.

```
 * @param args A map that must contain a 'config' key that points
 * to a parsed config (see readConfig()). Optionally, a
 * 'key' key can be provided which can be used to create a unique
 * name for the workflow process.
 *
 * @return A workflow that allows processing a channel of Vdsl3 formatted events
 * and apply a Viash config to them.
 ```

**Review by rcannood — CHANGES_REQUESTED** [review]
Minor changes

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 25 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf:834]: Yes, it assumes that `paramValues` contains only parameters that are in the viash config.

File: src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf (hunk 26 above)
Comment (DriesSchaumont) [reply src/main/resources/io/viash/platforms/nextflow/WorkflowHelper.nf]: Indeed, for several functions
